lib/bitcoin/BitcoinProcessor.ts

```diff
@@ -0,0 +1,581 @@
+          continue;
+        }
+        const data = Buffer.from(hexDataMatches[1], 'hex').toString();
+        if (data.startsWith(this.sidetreePrefix)) {
+          // we have found a sidetree transaction
+          const sidetreeTransaction: ITransaction = {
+            transactionNumber: TransactionNumber.construct(block, anchorFilePosition),
+            transactionTime: block,
+            transactionTimeHash: blockHash,
+            anchorFileHash: data.slice(this.sidetreePrefix.length)
+          };
+          console.debug(`Sidetree transaction found; adding ${JSON.stringify(sidetreeTransaction)}`);
+          anchorFilePosition++;
+          await this.transactionStore.addTransaction(sidetreeTransaction);
+        }
+      }
+    }
+    return blockHash;
+  }
+
+  /**
+   * performs an RPC call given a request
+   * @param request RPC request parameters as an object
+   * @param path optional path extension
+   * @returns response as an object
+   */
+  private async rpcCall (request: any, requestPath: string = ''): Promise<any> {
+    const fullPath = new URL(requestPath, this.bitcoinExtensionUri);
+    const requestString = JSON.stringify(request);
+    // console.debug(`Fetching ${fullPath}`);
+    // console.debug(requestString);
+    const response = await this.fetchWithRetry(fullPath.toString(), {
+      body: requestString,
+      method: 'post'
+    });
+
+    const responseData = await ReadableStreamUtils.readAll(response.body);
+    if (response.status !== httpStatus.OK) {
+      const error = new Error(`Fetch failed [${response.status}]: ${responseData}`);
+      console.error(error);
+      throw error;
+    }
+
+    const responseJson = JSON.parse(responseData);
+
+    if ('error' in responseJson && responseJson.error !== null) {
+      const error = new Error(`RPC failed: ${JSON.stringify(responseJson.error)}`);
+      console.error(error);
+      throw error;
+    }
+
+    return responseJson.result;
+  }
+
+  /**
+   * Calls node Fetch and retries the request on temporal errors
+   * @param uri URI to fetch
+   * @param requestParameters GET parameters to use
+   * @returns Response of the fetch
+   */
+  private async fetchWithRetry (uri: string, requestParameters?: RequestInit | undefined): Promise<Response> {
+    let retryCount = 0;
+    let timeout: number;
+    do {
+      timeout = this.defaultTimeout * 2 ** retryCount;
+      const params = Object.assign({}, requestParameters, {
+        timeout
+      });
+      try {
+        return await nodeFetch(uri, params);
+      } catch (error) {
+        if (error instanceof FetchError) {
+          if (retryCount >= this.maxRetries) {
+            console.debug('Max retries reached. Request failed.');
+            throw error;
+          }
+          switch (error.type) {
+            case 'request-timeout':
+              console.debug(`Request timeout (${retryCount})`);
+              await this.waitFor(Math.round(Math.random() * this.defaultTimeout + timeout));

```

Comment:
I don't get the intention behind `Math.random() * this.defaultTimeout`, it progressively less significant as retry increases, so can we remove it?

Reply:
I had put it in when there was greater parallelization, which is still an option in this code. 
In the event that multiple calls are made at the same time and the bitcoin peer becomes overwhelmed, the random would begin to separate the request retries. Its only one timeout interval because I wanted to retain some amount of guaranteed timeout between retries, but should still be significant enough to work over multiple retry attempts.